### ISSUE TITLE
Error on Windows with loading freetype?

### ISSUE THREAD

**Issue by astrofrog** [vispy/vispy#1276]
I'm having issues running e.g. the `axes_plot.py` example:

```
PS C:\Users\Administrator\vispy\examples\basics\scene> python .\axes_plot.py > windows.log
WARNING: Traceback (most recent call last):
  File ".\axes_plot.py", line 58, in <module>
    app.run()
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\app\_default_app.py", line 62, in run
    return default_app.run()
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\app\application.py", line 137, in run
    return self._backend._vispy_run()
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\app\backends\_qt.py", line 229, in _vispy_run
    return app.exec_()
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\app\backends\_qt.py", line 431, in event
    out = super(QtBaseCanvasBackend, self).event(ev)
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\app\backends\_qt.py", line 431, in event
    out = super(QtBaseCanvasBackend, self).event(ev)
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\app\backends\_qt.py", line 703, in paintGL
    self._vispy_canvas.events.draw(region=None)
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\util\event.py", line 455, in __call__
    self._invoke_callback(cb, event)
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\util\event.py", line 475, in _invoke_callback
    self, cb_event=(cb, event))
  << caught exception here: >>
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\util\event.py", line 471, in _invoke_callback
    cb(event)
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\scene\canvas.py", line 207, in on_draw
    self._draw_scene()
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\scene\canvas.py", line 253, in _draw_scene
    self.draw_visual(self.scene)
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\scene\canvas.py", line 291, in draw_visual
    node.draw()
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\scene\visuals.py", line 98, in draw
    self._visual_superclass.draw(self)
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\visuals\visual.py", line 588, in draw
    v.draw()
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\visuals\visual.py", line 432, in draw
    if self._prepare_draw(view=self) is False:
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\visuals\text\text.py", line 443, in _prepare_dra
w
    self._font._lowres_size) for t in text])
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\visuals\text\text.py", line 443, in <listcomp>
    self._font._lowres_size) for t in text])
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\visuals\text\text.py", line 157, in _text_to_vbo

    glyph = font[char]
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\visuals\text\text.py", line 69, in __getitem__
    self._load_char(char)
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\visuals\text\text.py", line 83, in _load_char
    _load_glyph(self._font, char, self._glyphs)
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\util\fonts\_freetype.py", line 47, in _load_glyp
h
    from ...ext.freetype import (FT_LOAD_RENDER, FT_LOAD_NO_HINTING,
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\ext\freetype.py", line 10, in <module>
    from ._bundled.freetype import *  # noqa
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\ext\_bundled\freetype.py", line 184, in <module>

    __dll__ = CDLL(FT_Library_filename)
  File "C:\Anaconda3\lib\ctypes\__init__.py", line 347, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: [WinError 126] The specified module could not be found
ERROR: Invoking <bound method SceneCanvas.on_draw of <SceneCanvas (PyQt4) at 0x6bd12d85f8>> for DrawEvent
ERROR: Invoking <bound method SceneCanvas.on_draw of <SceneCanvas (PyQt4) at 0x6bd12d85f8>> repeat 2
PS C:\Users\Administrator\vispy\examples\basics\scene> cat .\windows.log
PS C:\Users\Administrator\vispy\examples\basics\scene> Start-BitsTransfer-
PS C:\Users\Administrator\vispy\examples\basics\scene> Start-Transcript
Transcript started, output file is C:\Users\Administrator\Documents\PowerShell_transcript.WIN-S4OOT383BGP.NNT6Qtwu.2016
0926134428.txt
PS C:\Users\Administrator\vispy\examples\basics\scene> python .\axes_plot.py
WARNING: Traceback (most recent call last):
  File ".\axes_plot.py", line 58, in <module>
    app.run()
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\app\_default_app.py", line 62, in run
    return default_app.run()
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\app\application.py", line 137, in run
    return self._backend._vispy_run()
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\app\backends\_qt.py", line 229, in _vispy_run
    return app.exec_()
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\app\backends\_qt.py", line 431, in event
    out = super(QtBaseCanvasBackend, self).event(ev)
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\app\backends\_qt.py", line 431, in event
    out = super(QtBaseCanvasBackend, self).event(ev)
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\app\backends\_qt.py", line 703, in paintGL
    self._vispy_canvas.events.draw(region=None)
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\util\event.py", line 455, in __call__
    self._invoke_callback(cb, event)
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\util\event.py", line 475, in _invoke_callback
    self, cb_event=(cb, event))
  << caught exception here: >>
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\util\event.py", line 471, in _invoke_callback
    cb(event)
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\scene\canvas.py", line 207, in on_draw
    self._draw_scene()
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\scene\canvas.py", line 253, in _draw_scene
    self.draw_visual(self.scene)
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\scene\canvas.py", line 291, in draw_visual
    node.draw()
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\scene\visuals.py", line 98, in draw
    self._visual_superclass.draw(self)
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\visuals\visual.py", line 588, in draw
    v.draw()
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\visuals\visual.py", line 432, in draw
    if self._prepare_draw(view=self) is False:
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\visuals\text\text.py", line 443, in _prepare_dra
w
    self._font._lowres_size) for t in text])
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\visuals\text\text.py", line 443, in <listcomp>
    self._font._lowres_size) for t in text])
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\visuals\text\text.py", line 157, in _text_to_vbo

    glyph = font[char]
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\visuals\text\text.py", line 69, in __getitem__
    self._load_char(char)
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\visuals\text\text.py", line 83, in _load_char
    _load_glyph(self._font, char, self._glyphs)
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\util\fonts\_freetype.py", line 47, in _load_glyp
h
    from ...ext.freetype import (FT_LOAD_RENDER, FT_LOAD_NO_HINTING,
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\ext\freetype.py", line 10, in <module>
    from ._bundled.freetype import *  # noqa
  File "C:\Anaconda3\lib\site-packages\vispy-0.5.0.dev0-py3.5.egg\vispy\ext\_bundled\freetype.py", line 184, in <module>

    __dll__ = CDLL(FT_Library_filename)
  File "C:\Anaconda3\lib\ctypes\__init__.py", line 347, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: [WinError 126] The specified module could not be found
ERROR: Invoking <bound method SceneCanvas.on_draw of <SceneCanvas (PyQt4) at 0xf5fb1e95f8>> for DrawEvent
ERROR: Invoking <bound method SceneCanvas.on_draw of <SceneCanvas (PyQt4) at 0xf5fb1e95f8>> repeat 2
PS C:\Users\Administrator\vispy\examples\basics\scene> Stop-Transcript
Transcript stopped, output file is C:\Users\Administrator\Documents\PowerShell_transcript.WIN-S4OOT383BGP.NNT6Qtwu.2016
```

This appears to maybe be linked to Freetype?

**Comment by rougier** [comment]
I think the problem comes from ctypes taht cannot locate the freetype library (dll).

**Comment by astrofrog** [comment]
@rougier - I managed to fix this temporarily by removing the bundled freetype-py, installing the standalone package, and installing freetype with conda. However, I now have the following error:

```
WARNING: Traceback (most recent call last):
  File ".\axes_plot.py", line 58, in <module>
    app.run()
  File "c:\users\administrator\vispy\vispy\app\_default_app.py", line 62, in run
    return default_app.run()
  File "c:\users\administrator\vispy\vispy\app\application.py", line 137, in run
    return self._backend._vispy_run()
  File "c:\users\administrator\vispy\vispy\app\backends\_qt.py", line 229, in _vispy_run
    return app.exec_()
  File "c:\users\administrator\vispy\vispy\app\backends\_qt.py", line 431, in event
    out = super(QtBaseCanvasBackend, self).event(ev)
  File "c:\users\administrator\vispy\vispy\app\backends\_qt.py", line 431, in event
    out = super(QtBaseCanvasBackend, self).event(ev)
  File "c:\users\administrator\vispy\vispy\app\backends\_qt.py", line 703, in paintGL
    self._vispy_canvas.events.draw(region=None)
  File "c:\users\administrator\vispy\vispy\util\event.py", line 455, in __call__
    self._invoke_callback(cb, event)
  File "c:\users\administrator\vispy\vispy\util\event.py", line 475, in _invoke_callback
    self, cb_event=(cb, event))
  << caught exception here: >>
  File "c:\users\administrator\vispy\vispy\util\event.py", line 471, in _invoke_callback
    cb(event)
  File "c:\users\administrator\vispy\vispy\scene\canvas.py", line 207, in on_draw
    self._draw_scene()
  File "c:\users\administrator\vispy\vispy\scene\canvas.py", line 253, in _draw_scene
    self.draw_visual(self.scene)
  File "c:\users\administrator\vispy\vispy\scene\canvas.py", line 291, in draw_visual
    node.draw()
  File "c:\users\administrator\vispy\vispy\scene\visuals.py", line 98, in draw
    self._visual_superclass.draw(self)
  File "c:\users\administrator\vispy\vispy\visuals\visual.py", line 588, in draw
    v.draw()
  File "c:\users\administrator\vispy\vispy\visuals\visual.py", line 588, in draw
    v.draw()
  File "c:\users\administrator\vispy\vispy\visuals\visual.py", line 432, in draw
    if self._prepare_draw(view=self) is False:
  File "c:\users\administrator\vispy\vispy\visuals\text\text.py", line 443, in _prepare_draw
    self._font._lowres_size) for t in text])
  File "c:\users\administrator\vispy\vispy\visuals\text\text.py", line 443, in <listcomp>
    self._font._lowres_size) for t in text])
  File "c:\users\administrator\vispy\vispy\visuals\text\text.py", line 157, in _text_to_vbo
    glyph = font[char]
  File "c:\users\administrator\vispy\vispy\visuals\text\text.py", line 69, in __getitem__
    self._load_char(char)
  File "c:\users\administrator\vispy\vispy\visuals\text\text.py", line 102, in _load_char
    self._renderer.render_to_texture(data, self._atlas, (x, y), (w, h))
  File "c:\users\administrator\vispy\vispy\visuals\text\_sdf.py", line 272, in render_to_texture
    edf_neg_tex = self._render_edf(orig_tex)
  File "c:\users\administrator\vispy\vispy\visuals\text\_sdf.py", line 308, in _render_edf
    self.program_seed.draw('triangle_strip')
  File "c:\users\administrator\vispy\vispy\gloo\program.py", line 470, in draw
    canvas.context.flush_commands()
  File "c:\users\administrator\vispy\vispy\gloo\context.py", line 171, in flush_commands
    self.glir.flush(self.shared.parser)
  File "c:\users\administrator\vispy\vispy\gloo\glir.py", line 232, in flush
    self._shared.flush(parser)
  File "c:\users\administrator\vispy\vispy\gloo\glir.py", line 146, in flush
    parser.parse(self._filter(self.clear(), parser))
  File "c:\users\administrator\vispy\vispy\gloo\glir.py", line 458, in parse
    self._parse(command)
  File "c:\users\administrator\vispy\vispy\gloo\glir.py", line 420, in _parse
    ob.draw(*args)
  File "c:\users\administrator\vispy\vispy\gloo\glir.py", line 916, in draw
    gl.glDrawArrays(mode, first, count)
  File "c:\users\administrator\vispy\vispy\gloo\gl\_gl2.py", line 415, in glDrawArrays
    nativefunc(mode, first, count)
OSError: exception: access violation writing 0x0000003362351B94
```

Any idea what that could be due to?

**Comment by rougier** [comment]
Ouch, that is a severe one. Can you test the examples bundled with freetype-py to check if there's a problem with the bindings (https://github.com/rougier/freetype-py)

**Comment by mars0001** [comment]
Hi!
I have an issue that may be related to the one mentioned here.
Setup:
Windows 7 x64 (all updates) in VMWARE Environment with VmWare tools installed.
Python 3.6.3.0, vispy 0.6.0dev (tried with 0.5.3 same result), pyqt4, pyopengl, numpy

Tests are failing due of not finding the freetype dll:
`File "C:\DEV\WinPython3.6.3.0_x64\python-3.6.5.amd64\lib\site-packages\vispy\e
xt\freetype.py", line 10, in <module>
    from ._bundled.freetype import *  # noqa
  File "C:\DEV\WinPython3.6.3.0_x64\python-3.6.5.amd64\lib\site-packages\vispy\e
xt\_bundled\freetype.py", line 184, in <module>
    __dll__ = CDLL(FT_Library_filename)
  File "C:\DEV\WinPython3.6.3.0_x64\python-3.6.5.amd64\lib\ctypes\__init__.py",
line 348, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: [WinError 126] The specified module could not be found`

Is there a solution for this? Do I need some other software installed so Vispy tests can work?
BTW, nice work guys! I only suspect how much work you put on this. And passion and time ...

![vispy_issue](https://user-images.githubusercontent.com/40632281/41974332-58b0f4d2-7a20-11e8-96d7-1875cbf90f3b.JPG)

**Comment by rougier** [comment]
The latest freetype-py can be installed with  freetype binary, that could help.

**Comment by mars0001** [comment]
@rougier 
Hi Nicolas,
Can you guide me on how to do that?

I have: 
1. removed the "C:\DEV\WinPython3.6.3.0_x64\python-3.6.5.amd64\Lib\site-packages\vispy\ext\_bundled\freetype.py"
2. pip install freetype-py
3. downloaded the sources for freetype 2.9.1 from https://download.savannah.gnu.org/releases/freetype/

I work in Windows 10, should I compile the freetype sources? And then what do I need to do?

Thank you for any help you can give!

**Comment by mars0001** [comment]
Actually, for those who might get the same error, all is needed is to replace the file downloaded by VisPy named freetype253_x64.dll with the new freetype.DLL downloaded from https://github.com/ubawurinna/freetype-windows-binaries/tree/master/win64, renamed to freetype253_x64.dll.

**Comment by djhoese** [comment]
@larsoner Weren't you one of the people who set up the freetype stuff for vispy?

**Comment by larsoner** [comment]
Yep. Perhaps the DLL I compiled for vispy should be replaced with this other one.

**Comment by JoeXinfa** [comment]
I installed vispy using `pip install -e .` Now I get similar error. I'm in ubuntu and don't see the dll file. Could you help me out? Thanks in advance. @djhoese @larsoner 
```
joe@lenovo:~/code/vispy$ python examples/basics/plotting/scatter_histogram.py
WARNING: Traceback (most recent call last):
  File "examples/basics/plotting/scatter_histogram.py", line 40, in <module>
    fig.show(run=True)
  File "/home/joe/code/vispy/vispy/app/canvas.py", line 431, in show
    self.app.run()
  File "/home/joe/code/vispy/vispy/app/application.py", line 152, in run
    return self._backend._vispy_run()
  File "/home/joe/code/vispy/vispy/app/backends/_qt.py", line 237, in _vispy_run
    return app.exec_()
  File "/home/joe/code/vispy/vispy/app/backends/_qt.py", line 448, in event
    out = super(QtBaseCanvasBackend, self).event(ev)
  File "/home/joe/code/vispy/vispy/app/backends/_qt.py", line 448, in event
    out = super(QtBaseCanvasBackend, self).event(ev)
  File "/home/joe/code/vispy/vispy/app/backends/_qt.py", line 720, in paintGL
    self._vispy_canvas.events.draw(region=None)
  File "/home/joe/code/vispy/vispy/util/event.py", line 455, in __call__
    self._invoke_callback(cb, event)
  File "/home/joe/code/vispy/vispy/util/event.py", line 475, in _invoke_callback
    self, cb_event=(cb, event))
  << caught exception here: >>
  File "/home/joe/code/vispy/vispy/util/event.py", line 471, in _invoke_callback
    cb(event)
  File "/home/joe/code/vispy/vispy/scene/canvas.py", line 217, in on_draw
    self._draw_scene()
  File "/home/joe/code/vispy/vispy/scene/canvas.py", line 266, in _draw_scene
    self.draw_visual(self.scene)
  File "/home/joe/code/vispy/vispy/scene/canvas.py", line 304, in draw_visual
    node.draw()
  File "/home/joe/code/vispy/vispy/scene/visuals.py", line 99, in draw
    self._visual_superclass.draw(self)
  File "/home/joe/code/vispy/vispy/visuals/visual.py", line 588, in draw
    v.draw()
  File "/home/joe/code/vispy/vispy/visuals/visual.py", line 588, in draw
    v.draw()
  File "/home/joe/code/vispy/vispy/visuals/visual.py", line 432, in draw
    if self._prepare_draw(view=self) is False:
  File "/home/joe/code/vispy/vispy/visuals/text/text.py", line 521, in _prepare_draw
    self._font._lowres_size) for t in text])
  File "/home/joe/code/vispy/vispy/visuals/text/text.py", line 521, in <listcomp>
    self._font._lowres_size) for t in text])
  File "/home/joe/code/vispy/vispy/visuals/text/text.py", line 178, in _text_to_vbo
    glyph = font[char]
  File "/home/joe/code/vispy/vispy/visuals/text/text.py", line 81, in __getitem__
    self._load_char(char)
  File "/home/joe/code/vispy/vispy/visuals/text/text.py", line 95, in _load_char
    _load_glyph(self._font, char, self._glyphs)
  File "/home/joe/code/vispy/vispy/util/fonts/_freetype.py", line 50, in _load_glyph
    face = _load_font(f['face'], f['bold'], f['italic'])
  File "/home/joe/code/vispy/vispy/util/fonts/_freetype.py", line 37, in _load_font
    font = Face(fname)
  File "/home/joe/code/vispy/vispy/ext/_bundled/freetype.py", line 381, in __init__
    raise RuntimeError(hex(error))
RuntimeError: 0x1
ERROR: Invoking <bound method SceneCanvas.on_draw of <Fig (PyQt4) at 0x7fb1120d2240>> for DrawEvent
ERROR: Invoking <bound method SceneCanvas.on_draw of <Fig (PyQt4) at 0x7fb1120d2240>> repeat 2
```

**Comment by djhoese** [comment]
@JoeXinfa Do you have freetype installed? If so where are the library files?

**Comment by JoeXinfa** [comment]
Thanks David. I had it installed by conda, here `/home/joe/anaconda3/lib/libfreetype.so.6.16.1`

I looked at the file raised the error, `/home/joe/code/vispy/vispy/ext/_bundled/freetype.py`, and inserted `print(filename)` before the error. I'm surprised that it printed `/home/joe/Documents/demo-data/fonts/OpenSans-Regular.ttf`, which origins from the `vispy/util/fetching.py` in which I inserted the two lines below (to load from local disk instead of internet when I'm behind firewall). Then `git clone https://github.com/vispy/demo-data.git` and put there. It finally went through.
```
    dataDir = '/home/joe/Documents/demo-data'
    return os.path.join(dataDir, fname)
```
If I had not changed the `vispy/util/fetching.py`, it would do this:
```
Downloading data from http://github.com/vispy/demo-data/raw/master/fonts/OpenSans-Regular.ttf (212 kB)
[........................................] 100.00000 \ downloading   
File saved as /home/joe/.vispy/data/fonts/OpenSans-Regular.ttf.
```
If I had removed `~/.vispy`, it would do the above download/save again. So it seems the file `OpenSans-Regular.ttf` is a must-have, and vispy works hard to make it available. Then, to me, it makes more sense to have it inside the source package, but the team may have other considerations...

Lastly, I hope the original error message could be more clear.
